Documentation/gettingstarted/ciliumendpointslice.rst

```diff
@@ -0,0 +1,95 @@
+propagation should be reduced in this case, allowing for better scalability,
+at the cost of potentially longer delay before identity of new endpoint is
+recognized throughout the cluster.
+
+Deploy Cilium with CES
+=======================
+The CES feature relies on use of CEP. This feature is disabled by default
+in current beta status. You can enable the feature by setting the
+``enableCiliumEndpointSlice`` value to ``true``.
+
+.. parsed-literal::
+
+  helm install cilium |CHART_RELEASE| \\
+    --set enableCiliumEndpointSlice=true
+
+Verify that Cilium agent and operator pods are running.
+
+.. code-block:: shell-session
+
+   kubectl get pods -n kube-system -l k8s-app=cilium
+   NAMESPACE     NAME           READY   STATUS    RESTARTS   AGE
+   kube-system   cilium-7tnkq   1/1     Running   0          1d
+
+   kubectl get pods -n kube-system -l io.cilium/app=operator
+   NAMESPACE     NAME                               READY   STATUS    RESTARTS   AGE
+   kube-system   cilium-operator-855979b54b-w8rsm   1/1     Running   0          1d
+
+Validate that the CiliumEndpointSlice CRD has been registered.
+
+.. code-block:: shell-session
+
+   kubectl get crds --all-namespaces
+   NAME                                         CREATED AT
+   [...]
+   ciliumendpointslices.cilium.io               2021-11-05T05:41:28Z
+
+Known Issues and Workarounds
+============================
+
+Potential Racing Condition when Identity of An Existing Endpoint Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When there's an identity change for any existing resources, e.g., a pod,
+a deployment or a statefulset, although unlikely, the endpoints that
+undergo this change might experience connection disruption.
+
+Root cause for this potential disruption is that when identity of CEPs
+change, the operator will try to re-group/re-batch them into a different
+set of CESs. This breaks the atomic operation of an UPGRADE into that of
+an DELETE and an ADD. If the agent gets the DELETE (from old CES) first,
+it will remove the corresponding CEP's information from the ipcache,
+resulting in traffic to/from said CEP with an UNKNOWN identity.
+
+In current implementation, we added a delay (default: 1s) before sending
+out the DELETE event. This should greatly reduce the probability of
+connection disruption in most cases. 
+
+Upgrade Impact
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When performing an upgrade from a non-CES mode to CES mode, the operator
+will need to process all existing CEPs in the cluster and create CESs for
+them. Depending on the load of the cluster, this could take some time.
+
+During this period, new endpoints might experience longer delay before
+their identities will propagate to remote nodes.
+
+A mitigation strategy is to upgrade the operator from non-CES to CES first.
+This will let it start creating CESs right away, while the agents are still
+watching CEP resources.
```

Comment:
> how long is long enough to wait after upgrading the operator?

There's really no tell in current implementation, but I imagine we could either let the operator output some signal, or better, we could implement a check in the agent to see if CES store has all the CESs required to represent current CEPs, and only then start watching CES.

> What are the concrete steps to ensure that this upgrade occurs in this order

I think going GA, what we can do is to employ a 2-step upgrade again here. We put the operator change into some future 1.11 ver and let it create CES by default. Then in 1.12 release, we toggle agent watch by default. If we do this, then maybe there's no point implementing the checks I said above...

Reply:
oh also took your suggestion, edited text :)

Reply:
> There's really no tell in current implementation, but I imagine we could either let the operator output some signal, or better, we could implement a check in the agent to see if CES store has all the CESs required to represent current CEPs, and only then start watching CES.

I think that regardless of what else we do, this kind of signal would be really nice to have for debugability. Assuming that folks did the upgrade correctly and everything is fine works alright if everything is operating correctly and there's no unexpected bugs or issues in particular environments. But introspecting the actual state at a point in time and being able to make assessments on the health of the migration can be super helpful.

> > What are the concrete steps to ensure that this upgrade occurs in this order
> 
> I think going GA, what we can do is to employ a 2-step upgrade again here. We put the operator change into some future 1.11 ver and let it create CES by default. Then in 1.12 release, we toggle agent watch by default. If we do this, then maybe there's no point implementing the checks I said above...

In principle I like the idea, but I would be a bit concerned about silently turning this on in a future 1.11.x bugfix release. Enabling it in v1.11.0 and finding out that there is some issue with it (eg leaking resources, GC doesn't work, too high CPU, unexpected apiserver interactions, whatever the bugs may be) is not too bad. Introducing that later in the release cycle would increase the risk for upgrades that users will otherwise expect to be low-risk.

Reply:
> I think that regardless of what else we do, this kind of signal would be really nice to have for debugability.

I agree, created https://github.com/cilium/cilium/issues/17850 as a follow up item.

> Enabling it in v1.11.0 and finding out that there is some issue with it (eg leaking resources, GC doesn't work, too high CPU, unexpected apiserver interactions, whatever the bugs may be) is not too bad. 

Makes sense, should we enable operator piece by default now then (maybe in a quick follow up PR before 1.11 cut)?

Reply:
yeah please do follow up with a dedicated PR after we merge the core implementation and we can have that discussion / gauge the benefit vs. risk.